File: backend/service/leave.go
package service

type LeaveService interface {
	AddLeave(AddLeaveRequest, uint) (*LeaveResponse, error)
	GetLeavesByUser(uint) (*LeaveResponseWithCount, error)
	GetLeaves() ([]LeaveResponse, error)
	ApproveLeave(uint) error
	RejectLeave(uint) error
}

type AddLeaveRequest struct {
	Type      string `json:"type"`
	Detail    string `json:"detail"`
	TimeStart string `json:"time_start"`
	TimeEnd   string `json:"time_end"`
}

type LeaveResponse struct {
	ID        uint   `json:"id"`
	UserID    uint   `json:"user_id"`
	Type      string `json:"type"`
	Status    string `json:"status"`
	Detail    string `json:"detail"`
	TimeStart string `json:"time_start"`
	TimeEnd   string `json:"time_end"`
	CreatedAt string `json:"created_at"`
	UpdatedAt string `json:"updated_at"`
	LeaveDay  uint   `json:"leave_day"`
}

type LeaveResponseWithCount struct {
	LeaveResponse []LeaveResponse `json:"leave_responses"`
	CountSick     uint            `json:"count_sick"`
	CountBusiness uint            `json:"count_business"`
	CountVacation uint            `json:"count_vacation"`
}

File: frontend/store/leave.ts
import { createSlice } from "@reduxjs/toolkit";

export interface LeaveDTO {
  id: number;
  type: string;
  firstName: string;
  status: "requested" | "approved" | "rejected";
  time_start: string;
  time_end: string;
  detail: string;
  created_at: string;
  updated_at: string;
  leave_day: number;
}
export interface LeaveReponseWithCountDTO {
  leave: LeaveDTO[];
  count_sick: number;
  count_business: number;
  count_vacation: number;
}

export interface LeaveState {
  leave: LeaveDTO[]; // Change from LeaveDTO[] | null to always be an array
}

const leaveSlice = createSlice({
  name: "leave",
  initialState: {
    leave: [], // Initialize as an empty array, not null
  } as LeaveState,
  reducers: {
    setLeave: (state, action) => {
      state.leave = action.payload || []; // Ensure payload is an array
    },
    addLeave: (state, action) => {
      state.leave.push(action.payload);
    },
  },
});

export const { setLeave, addLeave } = leaveSlice.actions;
export default leaveSlice.reducer;
